helper.py

```diff
@@ -1,41 +1,119 @@
+=> Install asynio 0.2.0 via pip whl
 
-import os
-import time
+I will also update sheep to consume the websocket at the other end.
+'''
 
+import asyncio
+import aionotify
+from websockets.server import serve
+import websockets
+import time
+import base64
+import os
+from io import BytesIO
 from PIL import Image
 
 
-static_path = 'shepherd/static/'
+CONNECTIONS = set()
+
+#The following sets up the asynchronous waiting for file change
+watcher = aionotify.Watcher()
+
+#TODO - Change the input_file to the location of the image the camera takes!
+static_path = '/home/pi/shepherd/shepherd/static/'
 input_file = static_path + "full_image.jpg"
 output_file = static_path + "image.jpg"
 stat_file = static_path + "imgtime.txt"
 mod_time = 0
 out_width = 640.0
 out_height = 480.0
 
-while True:
-    time.sleep(0.1)
-    if os.path.exists(input_file):
-        stat_obj = os.stat(input_file)
-        if mod_time != stat_obj.st_mtime:
-            try:
-                img = Image.open(input_file)
-                width_i, height_i = img.size
-                width = width_i * 1.0
-                height = height_i * 1.0
-                scale_x = out_width / width
-                scale_y = out_height / height
-                if scale_x > scale_y:
-                    img = img.resize((int(width*scale_y), int(height*scale_y)), Image.ANTIALIAS)
-                else:
-                    img = img.resize((int(width*scale_x), int(height*scale_x)), Image.ANTIALIAS)
-
-                mod_time = stat_obj.st_mtime
-                img.save(output_file,"JPEG")
-                with open(stat_file, "w") as myfile:
-                    myfile.write(str(mod_time))
-            except IOError:
-                pass
+watcher.watch(alias='image', path=input_file, flags=aionotify.Flags.MODIFY)#sets up watcher
+
+def im_2_b64(image):
+    buff = BytesIO()
+    image.save(buff, format="JPEG")
+    img_str = base64.b64encode(buff.getvalue())
+    return img_str
+
+async def wait_for_file_change():
+    print("Waiting for file change")
+    loop = asyncio.get_event_loop()
+    c = 0
+    bypass = False#so that the very first image isn't ignored.
+    
+    try:#the following fails if full_image.jpg doesn't exist. This is VERY UNUSUAL but can happen.
+        await watcher.setup(loop)
+    except OSError:
+        print("Waiting for full_image.jpg")
+        while not os.path.exists(input_file):
+            time.sleep(0.5)#twiddle thumbs :)
+        await watcher.setup(loop)
+        bypass = True
+    
+    for _i in range(20):#for the first 20 events
+        if not bypass:
+            event = await watcher.get_event()#blocks until file changed
+        else:
+            bypass = False#reset bypass
+        print("Processing...")
+        failing = True
+        c = 0
+        while failing:
+            if c < 5:
+                try:#this runs until the bot has finished writing the image
+                    img = Image.open(input_file)
```

Comment:
So sorry to be like this, but I am pretty sure that `Image.open` does indeed open the image, because that's where the errors came from when the image was only half written. Also, it is the `image.open` command that creates the image object with the data of the image inside?

Reply:
In the [pillow docs](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.open)
> This is a lazy operation; this function identifies the file, but the file remains open and the actual image data is not read from the file until you try to process the data (or call the [load()](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image.load) method). See [new()](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.new). See [File Handling in Pillow](https://pillow.readthedocs.io/en/stable/reference/open_files.html#file-handling).

Sorry if I wasn't clear with my initial comment, but it only read basic information about the image and does some checks, waiting until an operation is performed to read the actual image data.